lib/utils/report.mjs

```diff
@@ -15,13 +16,46 @@ import { RULE_NAME_ALL } from '../constants.mjs';
  * @type {import('stylelint').Utils['report']}
  */
 export default function report(problem) {
-	const { ruleName, result, message, messageArgs, line, node, index, endIndex, word, severity } =
-		problem;
+	const {
+		index,
+		endIndex,
+		start,
+		end,
+		fix,
+		line,
+		message,
+		messageArgs,
+		node,
+		result,
+		result: {
+			stylelint: { config = {} },
+		},
+		ruleName,
+		severity,
+		word,
+	} = problem;
+
+	const hasRange = end?.column && start;
+	const hasIndexes = isNumber(index) && isNumber(endIndex);
+	const hasRangeData = hasRange || hasIndexes;
+	const hasFixCallback = typeof fix === 'function';
+	const hasFix = config.fix && !config.rules?.[ruleName][1]?.disableFix;
+	const hasFixData = hasFix && hasRangeData;
+
+	if (hasFixCallback && !hasRangeData) {
+		throw new Error(
+			`The fix callback for rule "${ruleName}" requires either index/endIndex or start/end to be passed to the \`report()\` function.`,
+		);
+	}
 
-	const ruleSeverityOption = severity || result.stylelint.ruleSeverities[ruleName];
+	if (hasFixData) {
```

Comment:
note

I chose not to expose the `FixerData` in the case of `context.fix` usage. It wouldn't incentivize enough to move towards using a fix callback.
This decision can be changed later on.